[TypeChecker] Incremental multi-statement closure type-checking (disabled by default)

xedin opened the pull request:
This is an attempt no improve multi-statement closure handling in the type-checker
to provide consistent semantics (expect to `inout` inference), better diagnostics and 
code completion for multi-statement  closures (including result builders).

There were attempts to improve this behavior over the years, the latest one being 
https://github.com/apple/swift/pull/32223 by Doug Gregor. All of them ran into technical 
difficulties related to internal organization of the type-checker, because multi-statement 
closures, just like function/subscript/setter bodies, tend be composed of a substantial 
number of statements and cannot be efficiently type-checked the same way as 
single-expression closures are. This proposal is backed by a new approach that uses 
incremental type-checking - each statement in the body is split into a number of independently 
type-checkable units - e.g. expression, declaration, pattern binding etc., these 
self-contained units are type-checked independently and inferred types are recorded
 to be used for type-checking of subsequent units. Because inference is done unit-by-unit 
and all unnecessary intermediate information is disregarded, this scheme is as efficient as 
approach used to type-check e.g. function bodies.

- New constraints:
  - Conjunction - represents a logical `and` relationship between its nested constraints, if one of the elements fails to produce a solution the conjunction as a whole is considered failed. There are two types of conjunctions - isolated and non-isolated. "Isolated" means that each of the elements has to be solved without any outside context unless it explicitly references a type variable e.g. anonymous closure parameter, capture or a closure result if element is a closure body node. 
  
  - Closure body element - represents a single AST node found in the body of the closure e.g. statement, declaration, or expression. Simplification of this constraint triggers constraint generation for the AST node. Although this constraint doesn't have any direct types it is allowed to references type variables e.g. if AST refers to some declarations found in other nodes.

- New step - `ConjunctionStep` - used to solve conjunction constraints, similar in that regard to existing `DisjunctionStep`, but has different semantics, namely, all of its elements have to produce a single solution for step to be considered successful;
   - Upon exhaustion of isolated conjunction, step returns all of the omitted external constraints back to the system and attempts to continue solving.

- Constraint generation for multi-statement closures happens incrementally across and inside of each of the statements in the body by means of generating a conjunction for each top-level and sub-elements, such conjunctions would have aforementioned `closure body element`s as its nested constraints that easier expanded in a new (smaller) conjunction (if it's a statement) or a set of solvable constraints (for declarations and expressions). This is helpful for diagnostics as well as performance because fail of a current element means omission of the rest.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[AST] Add a flag to guard exprimental multi-statement closures feature

xedin committed:
[ConstraintSystem] Add a skeleton of `ClosureBodyElement` constraint

xedin committed:
[ConstraintSystem] Add a skeleton of `Conjunction` constraint

It's similar to disjunction constraint but represents an "and"
relationship between its elements instead of "or", so all of the
elements have to produce a solution for conjunction constraint
to be considered solved successfully.

xedin committed:
[ConstraintSystem] Add a conjunction element producer

In preporation to add `ConjunctionStep`, let's first
introduce a `ConjunctionElement` and `ConjunctionElementProducer`.

xedin committed:
[CSStep] Add an implementation of `ConjunctionStep`

It behavies similar to `DisjunctionStep` but attempts all of
its elements unless there is an inference failure.

xedin committed:
[ConstraintSystem] Allow conjunction constraints to reference other variables

xedin committed:
[ConstraintSystem] Don't require a type variable per closure body element

Since each of the body elements is already going to have a type associated
with it there is no need to create yet another type variable.

xedin committed:
[ConstraintSystem] Allow `closure body element` constraint to reference type variables

This would be useful when once it's used as part of a conjunction to
be able to solve the closure body element component in isolation.

xedin committed:
[CSSimplify] Implement conjunction element simplification

xedin committed:
[ConstraintSystem] Implement new required methods in `ConjunctionElementProducer`

xedin committed:
[ConstraintSystem] Allow conjunctions marking conjunctions as isolated from the rest of the system

xedin committed:
[CSSolver] Don't select disabled conjunctions

xedin committed:
[Constraint] Implement `gatherReferencedVariables` for closure body element

xedin committed:
[ASTNode] Add a new element - statement condition

xedin committed:
[ConstraintLocator] Add a new element to track closure body elements

xedin committed:
[TypeChecker] Integrate experimental multi-statement closures flag into `shouldTypeCheckInEnclosingExpression`

xedin committed:
[ConstraintSystem] Expose type variables referenced by conjunction element

xedin committed:
[ConstraintSystem] Implement conjunction step

Iterate over all of the elements one-by-one and make sure that
each results in a single solution, otherwise fail the conjunction step.

Once all of the elements are handled either stop or,
if conjunction step has been performed in isolation,
return all of the outer constraints back to the system
and attempt to solve for outer context - that should
produce one or more solutions for conjunction to be
considered successfully solved.

xedin committed:
[CSClosure] Initial support for multi-statement closures

Implement constraint generation for brace statements and
refactor closure contraint generation to accept multi-statement
closures.

xedin committed:
[CSClosure] Implement constraint generation for closure body elements

xedin commented:
It's still early but everything is set up to start working on statements.

xedin committed:
[CSStep] Fail conjunction if element attempt fails

xedin committed:
[AST] Allow to acquire a pointer to statement condition

This is more of a (temporary) hack because condition could be reset
but that never happens during constraint solving, so it's relatively safe.

xedin committed:
[ConstraintSystem/Closures] Add support for `if` statement

xedin committed:
[CSClosure] NFC: Simplify creation of non-isolated conjunctions for a set of `ASTNode`s

xedin committed:
[CSClosure] Mark empty brace statements as non-viable elements

Such statements do not require any inference so could be safely
omitted from conjunctions.

xedin committed:
[CSStep] Filter solutions after each conjunction step

`SplitterStep` doesn't filter solutions if it results in a single
component (which is does for conjunction elements), so the solutions
have to be filtered manually after each element is done.

xedin committed:
[CSStep] Always restore snapshot once isolated conjunction step is done

Turn `SolverSnapshot::restore` into a destructor to make sure that
constraints are always returned when optional is reset.

xedin committed:
[Constraint] Don't propagate variables from elements to conjunction

Elements have to be solved in isolation, so if conjunction were
to reference all of the variables referenced by its elements then
it would have to bring unrelated type variables into its scope.

xedin committed:
[CSClosure] Further simplify conjunction creation for ASTNode elements

xedin committed:
[CSStep] Conjunction: integrate isolation scope into snapshot

It helps to simply handling of outer constrants because they have
to be added to the constraint system before scope is created but
constraint graph have to get updated after to make sure that
incremental binding inference already knows about types inferred
from conjunction.

xedin committed:
[ConstraintSystem/Closures] Add support for `guard` statement

xedin committed:
[ConstraintSystem/Closures] Add support for `while` statement

xedin committed:
[ConstraintSystem/Closures] Add support for `do` statement

xedin committed:
[ConstraintSystem/Closures] Add support for `repeat .. while` statement

xedin committed:
[Constraint] Allow closure body elements to carry contextual information

xedin committed:
[ConstraintSystem] Let closure body element simplification take advantage of contextual information

xedin committed:
[ConstraintSystem/Closures] Add support for `fallthrough` statement

xedin committed:
[ConstraintSystem/Closures] Add support for `defer` statement

xedin committed:
[ConstraintSystem/Closures] Add support for `break` statement

xedin committed:
[ConstraintSystem/Closures] Add support for `continue` statement

xedin committed:
[ConstraintSystem/Closures] Add support for `#assert` statement

xedin committed:
[ConstraintSystem/Closures] Add support for `throw` statement

xedin committed:
[ConstraintSystem] Add a new contextual purpose - CTP_ForEachSequence

A contextual purpose for a sequence expression associated with
`for-in` statement, that decays into a `ConformsTo` constraint
to a `Sequence` or `AsyncSequence` protocol.

Note that CTP_ForEachSequence is almost identical to CTP_ForEachStmt
but the meaning of latter is overloaded, so to avoid breaking solution
targets I have decided to add a new purpose for now.

xedin committed:
[ConstraintSystem/Closures] Add support for `for-in` statement

xedin committed:
[ASTNode] Add a new element - case item

xedin committed:
[ConstraintSystem] Add a new contextual purpose - CTP_CaseStmt

xedin committed:
[PreCheck] Don't attempt to pre-check patterns

Patterns found in different statements of a multi-statement closure
are going to be resolved while closure is type-checked.

xedin committed:
[ConstraintSystem/Closures] Add support for `case` statement

xedin committed:
[ConstraintSystem/Closures] Add support for `switch` statement

xedin committed:
[ConstraintSystem/Closures] Add support for `do-catch` statement

xedin committed:
[ConstraintSystem/Closures] Allow `return` without result expression

xedin committed:
[PreCheck] Don't walk into any declaration besides `PatternBindingDecl`

All other declarations encountered in a body of a multi-statement
closure are handled during solution application via calling
`TypeChecker::typeCheckDecl` that would run its own pre-check.

xedin committed:
[ConstraintSystem/Closures] Ignore all unsupported declarations during constraint generation

Everything besides `PatternBindingDecl` is handled at solution application time.

xedin committed:
[CSStep] Isolated conjunctions can't see outer solutions

All of the previously deduced solutions should be hidden
until isolated conjunction is successfully solved.

xedin committed:
[ConstraintSystem] Allow default initialized patterns

Add support for pattern bindings with default initializable patterns
e.g. `let x: Int?`.

xedin commented:
@swift-ci please test source compatibility

xedin committed:
[CSClosure] Mark conjunctions representing closure body as isolated

Bodies of multi-statement closures should be handled in isolation
because they don't dependent on anything expect parameter/result
types from outer context.

xedin committed:
[AST] Allow pattern bindings to be marked as fully verified

This would be used by closures because they handle patterns
and initializers via solver and set them back to the pattern
binding decl as fully type-checked.

xedin committed:
[CSApply] Mark pattern target as fully verified

xedin committed:
[CSClosure] Allow all declarations to use `typeCheckDecl`

This is necessary to ensure that all of the information is
properly materialized, and access/availability and other
required checks are performed.

xedin committed:
[CSStep] Conjunctions representing closures affect declaration context

While solving a conjunction that represents a (multi-statement) closure
constraint system should use such closure as its declaration context,
otherwise member lookup would produce incorrect results.

xedin commented:
@swift-ci please test source compatibility

xedin committed:
[CSApply] Delay solution application to multi-statement closure bodies

Let's delay solution application to multi-statement closure bodies,
because declarations found in the body of such closures may depend
on its `ExtInfo` flags e.g. no-escape is set based on parameter if
closure is used in a call.

xedin committed:
[CSClosure] Account for partially resolved inner parameter types

It is possible that contextual type of a parameter
has been assigned to an anonymous of named argument
early, to facilitate closure type checking. Such a
type can have type variables inside e.g.

```swift
func test<T>(_: (UnsafePointer<T>) -> Void) {}

test { ptr in
  ...
}
```

Type variable representing `ptr` in the body of
this closure would be bound to `UnsafePointer<$T>`
in this case, where `$T` is a type variable for a
generic parameter `T`.

xedin committed:
[ResultBuilders] Fix a crash in `PreCheckResultBuilderApplication`

Cant assume that `shouldTypeCheckInEnclosingExpression` always
implies a single expression closure anymore.

xedin committed:
[ConstraintSystem] Compute variables referenced by conjunction elements incrementally

Attempting to pre-compute a set of referenced type variables
upfront is incorrect because parameter(s) and/or result type
could be bound before conjunction is attempted. Let's compute
a set of referenced variables before each element gets attempted.

xedin committed:
[CSClosure] Avoid creating conjunctions for empty bodies

xedin committed:
[CSClosure] Link closure conjunction with referenced outer parameters

If one or more outer closure parameters used in the body of an
inner closure have not been resolved when conjunction is created,
let's reference them in the conjunction constraint to avoid
disconnecting conjunction from its context.

xedin commented:
@swift-ci please test source compatibility

xedin committed:
[CSApply] Set types for for all delayed closures

xedin commented:
@swift-ci please test source compatibility

sarru reviewed (commented):
don't depend* on anything except* ( mistake in pull request description)

xedin committed:
[CSClosure] Avoid function conversion when closure result type is optional `Void`

In cases like:

```
func test<T>(_: () -> T?) -> [T] {
   ...
}

test {
  if ... {
     return
  }

  ...
}
```

Contextual return type would be first attempted as optional and
then unwrapped if the first attempt did not succeed. To avoid having
to solve the closure twice (which results in an implicit function conversion),
let's add an implicit empty tuple (`()`) to the `return` statement
and allow it be to injected into optional as many times as context
requires.

xedin committed:
[CSClosures] Model `return` statements as solution application targets

Each return statement gets a contextual type from outer context,
so the best model for that is via solution application target.

xedin committed:
[CSClosure] Handle pattern and sequence of for-in loop together

Let `visitForEachPattern` handle both pattern and the sequence
expression associated with `for-in` loop because types in this
situation flow in both directions:

- From pattern to sequence, informing its element type e.g.
  `for i: Int8 in 0 ..< 8`

- From sequence to pattern, when pattern has no type information.

xedin committed:
[CSClosure] Check parameters after solution is applied to the body of a closure

xedin committed:
[CSClosure] Make sure that `case` body variables get types

Solution application needs to set interface types for variables
declared in a case statement, otherwise this would lead to crashes
in SILGen.

xedin committed:
[CSClosure] Make sure that partially resolved params/result gets their type variables referenced

It's possible to declare explicit parameter/result type without
providing generic arguments, in such situation they'd be "opened"
as type variables and inferred from the contextual type or the body.

Type variable collector needs to account for that and include
such inner type variables into scope of an isolated closure.

hborla reviewed (commented):
I'm about half way through the commits, but here's a first round of questions. All of the added constraint generation and application for various statement kinds looks relatively straightforward to me, and the questions below are related to the new conjunction constraint infrastructure.

xedin commented on lib/Sema/CSStep.cpp:
```diff
@@ -354,6 +354,9 @@ StepResult ComponentStep::take(bool prevFailed) {
     // Produce a disjunction step.
     return suspend(
         std::make_unique<DisjunctionStep>(CS, disjunction, Solutions));
+  } else if (auto *conjunction = CS.selectConjunction()) {
```
It's just models the way how closures used to be type-checked before, wait until everything that can be resolved.

xedin commented on include/swift/Sema/Constraint.h:
```diff
@@ -780,6 +788,10 @@ class Constraint final : public llvm::ilist_node<Constraint>,
   bool isExplicitConversion() const;
 
+  /// Determine whether this constraint should be solved in isolation
+  /// from the rest of the constraint system.
```
One-way constraints form a separate component for left-hand side type variable, in contrast isolated closures have to remove all of the inactive constraints and unrelated type variables and preserve only type variables related to parameter and result, they can't be separated into a component like one-way are because all of the constraints around a closure would have to be brought back after the body is type-checked.

xedin commented on lib/Sema/CSStep.cpp:
```diff
@@ -783,10 +783,46 @@ bool DisjunctionStep::attempt(const DisjunctionChoice &choice) {
+    CS.addTypeVariable(typeVar);
+
+  // Make sure that element is solved in isolation
+  // by dropping all scoring information.
```
Because score from the outside is unrelated to the elements in the body, solutions can't be ranked on score from context and score from other elements because they are standalone units for type-checker, only types are transferable between them.

hborla commented on lib/Sema/CSClosure.cpp:
```diff
@@ -38,8 +38,36 @@ class TypeVariableRefFinder : public ASTWalker {
           referencedVars.insert(typeVar);
+
+          // It is possible that contextual type of a parameter
+          // has been assigned to an anonymous of named argument
```
Should this say "anonymous **or** named"?

hborla replied:
(resolved in a follow-up commit)

hborla reviewed (approved):
This looks really solid. I didn't notice anything suspicious in the remainder of the commits, and your explanatory comments are really helpful.

xedin commented:
@swift-ci please clean test

xedin commented:
@swift-ci please test source compatibility

xedin committed:
[CSClosure] NFC: Add size to a `SmallVector` in `ClosureConstraintGenerator::visitIfStmt`

xedin commented:
@swift-ci please clean test

xedin commented:
@swift-ci please test source compatibility

xedin committed:
[CSClosure] Check multi-statement closure attrs only in if inference is enabled

xedin committed:
[PreCheck] Allow pre-check to walk into patterns when multi-statement inference is disabled

This restores original pre-check behavior and is important for diagnostics
like `'_' can only appear in a pattern or on the left side of an assignment`.

xedin commented:
@swift-ci please test source compatibility release

xedin commented:
There is C++ interop related test-case failure in watchsimulator, going to run smoke test instead and land this.

xedin commented:
@swift-ci please test Windows platform

xedin commented:
@swift-ci please test macOS platform
